tests/Unit/DataStructures/DataBox/Test_ObservationBox.cpp

```diff
@@ -0,0 +1,128 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#include "Framework/TestingFramework.hpp"
+
+#include "DataStructures/DataBox/DataBox.hpp"
+#include "DataStructures/DataBox/ObservationBox.hpp"
+#include "DataStructures/DataBox/Tag.hpp"
+#include "DataStructures/DataVector.hpp"
+#include "DataStructures/Tensor/Tensor.hpp"
+#include "DataStructures/Variables.hpp"
+#include "DataStructures/VariablesTag.hpp"
+#include "Utilities/Gsl.hpp"
+#include "Utilities/TMPL.hpp"
+
+namespace {
+void multiply_by_two(const gsl::not_null<double*> result, const double value) {
+  *result = 2.0 * value;
+}
+
+void add_all(const gsl::not_null<double*> result, const double value0,
+             const double value1, const double value2) {
+  *result = value0 + value1 + value2;
+}
+
+struct Tag0 : db::SimpleTag {
+  using type = double;
+};
+
+struct Tag1 : db::SimpleTag {
+  using type = double;
+};
+
+struct Tag1Compute : Tag1, db::ComputeTag {
+  using base = Tag1;
+  using return_type = double;
+  static constexpr auto function = multiply_by_two;
+  using argument_tags = tmpl::list<Tag0>;
+};
+
+struct Tag2 : db::SimpleTag {
+  using type = double;
+};
+
+struct Tag2Compute : Tag2, db::ComputeTag {
+  using base = Tag2;
+  using return_type = double;
+  static constexpr auto function = multiply_by_two;
+  using argument_tags = tmpl::list<Tag1>;
+};
+
+struct Tag3 : db::SimpleTag {
+  using type = double;
+};
+
+struct Tag3Compute : Tag3, db::ComputeTag {
+  using base = Tag3;
+  using return_type = double;
+  static constexpr auto function = add_all;
+  using argument_tags = tmpl::list<Tag0, Tag1, Tag2>;
+};
+
+struct ScalarVar : db::SimpleTag {
+  using type = Scalar<DataVector>;
+};
+
+struct VectorVar : db::SimpleTag {
+  using type = tnsr::I<DataVector, 3, Frame::Inertial>;
+};
+
+struct VarsCompute : db::ComputeTag,
+                     ::Tags::Variables<tmpl::list<ScalarVar, VectorVar>> {
+  using base = ::Tags::Variables<tmpl::list<ScalarVar, VectorVar>>;
+  using return_type = typename base::type;
+  using argument_tags = tmpl::list<>;
+  static void function(
+      const gsl::not_null<::Variables<tmpl::list<ScalarVar, VectorVar>>*>
+          result) {
+    result->initialize(5, 1.0);
+    for (size_t i = 0; i < 3; ++i) {
+      get<VectorVar>(*result).get(i) += static_cast<double>(i) + 1.0;
+    }
+  }
+};
+
+// Test that we can return values from the apply functions and that we can
+// retrieve both the nested DataBox and the ObservationBox via the argument
+// tags.
+struct SubtractNumberAndReturn {
+  using argument_tags =
+      tmpl::list<::Tags::ObservationBox, ::Tags::DataBox, Tag3>;
+
+  template <typename DataBoxType, typename ComputeTagsList>
+  double operator()(const ObservationBox<ComputeTagsList, DataBoxType>& obs_box,
+                    const DataBoxType& box, const double tag3,
```

Comment:
Perhaps write `DataBoxType` -> `db::DataBox<DbTagsList>` to make sure the type is a DataBox here?